inc/namespace.php

```diff
@@ -356,3 +389,86 @@ function add_fallback_sizes( array $metadata, int $attachment_id ) : array {
 
 	return $metadata;
 }
+
+function get_valid_dimensions( $size, int $max_width = 0, int $max_height = 0 ) : array {
+	$width = $max_width;
+	$height = $max_height;
+	$crop = false;
+
+	if ( is_array( $size ) ) {
+		$width = $size[0];
+		$height = $size[1];
+		$crop = $size[2] ?? $crop;
+	} else {
+		$all_sizes = wp_get_registered_image_subsizes();
+		if ( ! isset( $all_sizes[ $size ] ) ) {
+			return [ $width, $height, $crop ];
+		}
+
+		$width = $all_sizes[ $size ]['width'];
+		$height = $all_sizes[ $size ]['height'];
+		$crop = $all_sizes[ $size ]['crop'];
+	}
+
+	if ( $crop ) {
+		return array_merge(
+			wp_constrain_dimensions( $width, $height, $max_width, $max_height ),
+			[ $crop ]
+		);
+	}
+
+	$width = min( $width, $max_width );
+	$height = min( $height, $max_height );
+
+	return [ $width, $height, $crop ];
+}
+
+function dynamic_downsize( $downsize, int $attachment_id, $size ) {
+	$attachment = get_post( $attachment_id );
+
+	$provider = get_asset_provider( $attachment );
+	if ( empty( $provider ) || ! $provider instanceof Resizable ) {
```

Comment:
`instanceof` can take `null` as left operand, so this could also just be this:

```suggestion
	if ( ! $provider instanceof Resizable ) {
```

Same further down.

Reply:
Ahh cool, I think this is also because previously it was using the `$provider->support_dynamic_image_sizing()` method.